src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -406,34 +402,131 @@ public SymbolTable getSymbolTable()
+        } else if (this instanceof IonContainerLite) {
+            return containerHashCode();
+        }
+        return scalarHashCode();
     }
 
-    /*
-     * Internal HashCode implementation which utilizes a SymbolTableProvider
-     * to resolve the SymbolTable to use in encoding the sub-graph.
+    private int containerHashCode() {
+        HashHolder[] hashStack = new HashHolder[CONTAINER_STACK_INITIAL_CAPACITY];
+        int hashStackIndex = 0;
+        hashStack[hashStackIndex] = new HashHolder();
+        IonValueLite value = this;
+        do {
+            HashHolder hashHolder = hashStack[hashStackIndex];
+            if ((value._flags & IS_NULL_VALUE) != 0) {
+                // Null values are hashed using a constant signature unique to the type.
+                hashHolder.update(value.hashTypeAnnotations(value.hashSignature()), value);
+            } else if (!(value instanceof IonContainerLite)) {
+                hashHolder.update(value.scalarHashCode(), value);
+            } else {
+                // Step into the container by pushing a HashHolder for the container onto the stack.
+                if (++hashStackIndex >= hashStack.length) {
+                    hashStack = growHashStack(hashStack);
+                }
+                hashHolder = hashStack[hashStackIndex];
+                if (hashHolder == null) {
+                    hashHolder = new HashHolder();
+                    hashStack[hashStackIndex] = hashHolder;
+                }
+                hashHolder.parent = (IonContainerLite) value;
+                hashHolder.iterator = hashHolder.parent.new SequenceContentIterator(0, true);
+                hashHolder.valueHash = value.hashSignature();
+            }
+            do {
+                if (hashHolder.parent == null) {
+                    // Iteration has returned to the top level; return the top-level value's hash code.
+                    return hashHolder.valueHash;
+                }
+                value = hashHolder.iterator.nextOrNull();
+                if (value == null) {
+                    // The end of the container has been reached. Pop from the stack and update the parent's hash.
+                    hashHolder = hashStack[hashStackIndex--];
+                    IonValueLite container = hashHolder.parent;
+                    int containerHash = container.hashTypeAnnotations(hashHolder.valueHash);
+                    hashHolder.parent = null;
+                    hashHolder.iterator = null;
+                    hashHolder = hashStack[hashStackIndex];
+                    hashHolder.update(containerHash, container);
+                }
+            } while (value == null);
+        } while (true);
+    }
+
+    private static HashHolder[] growHashStack(HashHolder[] hashStack) {
+        HashHolder[] newHashStack = new HashHolder[hashStack.length * 2];
+        System.arraycopy(hashStack, 0, newHashStack, 0, hashStack.length);
+        return newHashStack;
```

Comment:
```suggestion
        return Arrays.copyOf(hashStack, hashStack.length*2);
```

Same elsewhere, comment if not performance neutral.

Reply:
Done in #562